status-server/src/public/js/client.js

```diff
@@ -1,199 +1,168 @@
-  '/user/5/status': {},
-  '/user/6/status': {}
+/* global config */
+
+var appState = {
+  mySlot: {
+    id: config.id,
+    status: {}
+  },
+  slots: []
 };
 
-function objectValues(obj) {
-  return Object.keys(obj).map(key => obj[key]);
-}
+/*
+an example of the slots array
+[
+  null,
+  {
+    id: "5",
+    status: {
+      value: 1,
+      last-modified: ''
+    }
+  },
+  {id: "2", status: null},
+  {id: "6", status: {...}},
+  null,
+  {id: "3", status: {...}}
+]
+*/
 
-function startPolling() {
-  stopPolling();
-  intervalID = setInterval(requestStatus, 2000);
-}
+function init() {
+  jsonRequest('GET', 'user/'+ config.id +'/status', function(err, data) {
+    if (err) {
+      console.error('Could not request my status');
+      return;
+    }
 
-function stopPolling() {
-  if (intervalID) {
-    clearInterval(intervalID);
-  }
-  intervalID = null;
-}
+    appState.mySlot.status = data;
 
-function togglePolling(btn) {
-  if (intervalID) {
-    stopPolling();
-    if(btn) {
-      btn.value = "Start";
-    }
-  } else {
-    startPolling();
-    if (btn) {
-      btn.value = "Stop";
-    }
-  }
+    renderApp();
+  });
+
+  getMySlots();
+
+  var intervalID = setInterval(getMySlots, 2000);
 }
 
-function init() {
-  // update the main heading if this client instance was passed an id
-  if (config) {
-    // assign a default id if none was passed
-    if (!config.id) {
-      config.id = 'user0';
-    }
-    var titleNode = document.querySelector('h1');
-    if (titleNode) {
-      titleNode.innerHTML = 'status client: ' + config.id;
+function getMySlots() {
+  jsonRequest('GET', 'user/'+ config.id +'/slots', function(err, data) {
+    if (err) {
+      console.error('Could not request my slots');
+      return;
     }
-    // temporarily map the user id to the led node just by aligning the indices.
-    // We'll want a proper slots=>users mapping eventually
-    Array.prototype.forEach.call(document.querySelectorAll('.led'), (node, idx) => {
-      var userId = node.id.replace(/^led([0-9])+/, 'user$1');
-      node.dataset.user = userId;
-      if (config && config.id.endsWith(idx)) {
-        node.classList.add('isSelf');
+
+    appState.slots = data.value.map(function(id){
+      if (id === null) {
+        return null;
       }
+      
+      return {
+        id: id,
+        status: null
+      };
     });
-  }
 
-  if (config && config.autostart) {
-    togglePolling(document.getElementById('pollingBtn'));
-  }
+    var reqTotal = 0;
+    var reqComplete = 0;
+
+    appState.slots.forEach(function(slot) {
+      if (slot === null) {
+        return;
+      }
+
+      reqTotal += 1;
+
+      jsonRequest('GET', 'user/'+ slot.id +'/status', function(err, data) {
+        if (err) {
+          console.error('Could not request status of' + slot.id);
+          return;
+        }
+        slot.status = data;
+        
+        reqComplete += 1;
+        if (reqComplete === reqTotal) {
+          renderApp();
```

Comment:
Some slots would not be updated correctly if I did not block the rendering until all slots status' were fetched. It's very similar to using a promise, accumulating the results and waiting for them to resolve, but this is the callback version.